src/wallet/wallet.cpp

```diff
@@ -455,8 +457,10 @@ bool CWallet::IsSpent(const uint256& hash, unsigned int n) const
     {
         const uint256& wtxid = it->second;
         std::map<uint256, CWalletTx>::const_iterator mit = mapWallet.find(wtxid);
-        if (mit != mapWallet.end() && mit->second.GetDepthInMainChain() >= 0)
-            return true; // Spent
+        if (mit != mapWallet.end()) {
+            if (mit->second.GetDepthInMainChain() > 0  || (mit->second.GetDepthInMainChain() == 0 && !mit->second.isAbandoned()))
```

Comment:
`CMerkleTx::GetDepthInMainChain()` is slightly expansive (and `IsSpent()` gets called a lot). Would caching `mit->second.GetDepthInMainChain()` make sense? Or do we expect the compiler does it internally (I don't think so)?